pkg/tsdb/cloudmonitoring/cloudmonitoring.go

```diff
@@ -622,35 +633,76 @@ func (e *CloudMonitoringExecutor) parseResponse(queryRes *tsdb.QueryResult, data
 						// https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries#Distribution
 						bucketBound := calcBucketBound(point.Value.DistributionValue.BucketOptions, i)
 						additionalLabels := map[string]string{"bucket": bucketBound}
-						buckets[i] = &tsdb.TimeSeries{
-							Name:   formatLegendKeys(series.Metric.Type, defaultMetricName, nil, additionalLabels, query),
-							Points: make([]tsdb.TimePoint, 0),
+
+						timeField := data.NewField(data.TimeSeriesTimeFieldName, nil, []time.Time{})
+						valueField := data.NewField(data.TimeSeriesValueFieldName, nil, []float64{})
+
+						frameName := formatLegendKeys(series.Metric.Type, defaultMetricName, nil, additionalLabels, query)
+						valueField.Name = frameName
+						buckets[i] = &data.Frame{
+							Name: frameName,
+							Fields: []*data.Field{
+								timeField,
+								valueField,
+							},
+							RefID: query.RefID,
 						}
+
 						if maxKey < i {
 							maxKey = i
 						}
 					}
-					buckets[i].Points = append(buckets[i].Points, tsdb.NewTimePoint(null.FloatFrom(value), float64((point.Interval.EndTime).Unix())*1000))
+					buckets[i].AppendRow(point.Interval.EndTime, value)
 				}
-
-				// fill empty bucket
 				for i := 0; i < maxKey; i++ {
 					if _, ok := buckets[i]; !ok {
 						bucketBound := calcBucketBound(point.Value.DistributionValue.BucketOptions, i)
-						additionalLabels := map[string]string{"bucket": bucketBound}
-						buckets[i] = &tsdb.TimeSeries{
-							Name:   formatLegendKeys(series.Metric.Type, defaultMetricName, seriesLabels, additionalLabels, query),
-							Points: make([]tsdb.TimePoint, 0),
+						additionalLabels := data.Labels{"bucket": bucketBound}
+						timeField := data.NewField(data.TimeSeriesTimeFieldName, nil, []time.Time{})
+						valueField := data.NewField(data.TimeSeriesValueFieldName, nil, []float64{})
+						frameName := formatLegendKeys(series.Metric.Type, defaultMetricName, seriesLabels, additionalLabels, query)
+						valueField.Name = frameName
+						buckets[i] = &data.Frame{
+							Name: frameName,
+							Fields: []*data.Field{
+								timeField,
+								valueField,
+							},
+							RefID: query.RefID,
 						}
 					}
 				}
 			}
 			for i := 0; i < len(buckets); i++ {
-				queryRes.Series = append(queryRes.Series, buckets[i])
+				frames = append(frames, buckets[i])
 			}
 		}
 	}
 
+	if len(cmr.TimeSeries) > 0 {
+		dl := query.buildDeepLink()
+		for i := range frames {
+			for j, field := range frames[i].Fields {
+				if field.Config == nil {
+					frames[i].Fields[j].Config = &data.FieldConfig{}
+				}
+				deepLink := data.DataLink{
+					Title:       "View in Metrics Explorer",
+					TargetBlank: true,
+					URL:         dl,
+				}
+				frames[i].Fields[j].Config.Links = append(frames[i].Fields[j].Config.Links, deepLink)
```

Comment:
here is where i put the deeplink and unit which were the logics on the frontend part into backend dataframe.fields.config. I am not sure that we need to put it into all fields.config or just fields[1].config (just for the data not the time serie)

Reply:
That's a good question. I think it needs to be put on all fields.config just like you've done, but it would certainly be neater to put it fields[1].config. Can you try that and see if it's sufficient?

Reply:
I just updated the pr, after testing, it seems we need only to set the frames[i].Fields[1], it is enough.